File: medium/131. Palindrome Partitioning/131. Palindrome Partitioning.ts
function isPalindrome(s: string) {
    return s === s.split("").reverse().join("")
}

function partition(s: string): string[][] {
    const n = s.length
    const res: string[][] = []

    const backtrack = (index: number, path: string[]) => {
        if (index === n) {
            res.push(path)
            return
        }

        for (let i = index + 1; i <= n; i++) {
            if (isPalindrome(s.slice(index, i)))
                backtrack(i, [...path, s.slice(index, i)])
        }
    }

    backtrack(0, [])

    return res
};


export { }

File: medium/131. Palindrome Partitioning/131. Palindrome Partitioning.py
from typing import List


class Solution:
    def is_palindrome(self, s: str) -> bool:
        return s == s[::-1]

    def partition(self, s: str) -> List[List[str]]:
        n = len(s)
        res: List[List[str]] = []

        def backtrack(path: List[str], index: int):
            if index == n:
                res.append(path[:])
                return

            for i in range(index + 1, n + 1):
                if self.is_palindrome(s[index:i]):
                    backtrack(path + [s[index:i]], i)

        backtrack([], 0)

        return res
